Plugins/BuildServerIntegration/VstsAndTfsIntegration/TfsApiHelper.cs

```diff
@@ -0,0 +1,186 @@
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using GitUI;
+using Microsoft.VisualStudio.Threading;
+using Newtonsoft.Json;
+
+namespace VstsAndTfsIntegration
+{
+    /// <summary>
+    /// For VSTS and TFS >=2015
+    /// </summary>
+    public class TfsApiHelper : IDisposable
+    {
+        private HttpClient _httpClient;
+        private IEnumerable<BuildDefinition> _buildDefinitions;
+        private string _projectId;
+
+        private HttpClient InitializeHttpClient(string serverUrl, string personalAccessToken)
+        {
+            var client = new HttpClient();
+            client.DefaultRequestHeaders.Accept.Add(
+                new MediaTypeWithQualityHeaderValue("application/json"));
+
+            client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic",
+                Convert.ToBase64String(
+                    System.Text.Encoding.ASCII.GetBytes($":{personalAccessToken}")));
+
+            var formatedUrl = serverUrl.EndsWith("/") ? serverUrl : serverUrl + "/";
+            client.BaseAddress = new Uri(formatedUrl);
+            return client;
+        }
+
+        private async Task<IEnumerable<BuildDefinition>> GetBuildDefinitionsAsync(string teamCollection, string projectName, Regex buildDefinitionNameFilter)
+        {
+            var projects = await GetProjectsAsync(teamCollection);
+            var buildDefs = new List<BuildDefinition>();
+
+            if (projects.Count == 0)
+            {
+                return buildDefs;
+            }
+
+            var projectFound = projects.FirstOrDefault(p => p.Name == projectName);
+            if (projectFound == null)
+            {
+                return buildDefs;
+            }
+
+            _projectId = projectFound.Id;
+            using (var response = await _httpClient.GetAsync($"{_projectId}/_apis/build/definitions"))
+            {
+                response.EnsureSuccessStatusCode();
+                string json = await response.Content.ReadAsStringAsync();
+
+                var buildDefinitions = JsonConvert.DeserializeObject<ListWrapper<BuildDefinition>>(json);
+
+                if (buildDefinitionNameFilter == null)
+                {
+                    buildDefs.AddRange(buildDefinitions.Value);
+                }
+                else
+                {
+                    foreach (var buildDefinition in buildDefinitions.Value)
+                    {
+                        if (buildDefinitionNameFilter.IsMatch(buildDefinition.Name))
+                        {
+                            buildDefs.Add(buildDefinition);
+                        }
+                    }
+                }
+            }
+
+            return buildDefs;
+        }
+
+        private async Task<IList<Project>> GetProjectsAsync(string teamCollection)
+        {
+            using (var response = await _httpClient.GetAsync($"{teamCollection}/_apis/projects"))
```

Comment:
This HTTP call returns results up to the first 100 projects, and will fail to find projects after that first page. The API supports pagination though, so it should be easy to iterate over projects (and performance shouldn't suffer too much). Something like that:

```csharp
        private const int RequestedProjectsPerPage = 100;

        private async Task<IList<Project>> GetProjectsAsync(string teamCollection)
        {
            int nextPage = 0;
            IList<Project> allProjects = new List<Project>();
            ListWrapper<Project> currentProjects;
            do
            {
                currentProjects = await DoOneGetProjectsAsync(teamCollection, RequestedProjectsPerPage, nextPage);

                allProjects.AddAll(currentProjects.Value);
                nextPage += currentProjects.Value.Count;
            }
            while (currentProjects.Count == RequestedProjectPerPage);

            return allProjects;
        }

        private async Task<ListWrapper<Project>> DoOneGetProjectsAsync(string teamCollection, int top, int skip)
        {
            using (var response = await _httpClient.GetAsync($"{teamCollection}/_apis/projects?$top={top}&$skip={skip}"))
            {
                response.EnsureSuccessStatusCode();
                string json = await response.Content.ReadAsStringAsync();
 

                return JsonConvert.DeserializeObject<ListWrapper<Project>>(json);
            }
        }
```

Reply:
I have seen that we could use the project `name` in the uri instead of the `id` to query build definitions and builds( See `project | string | Team project ID or name` in 
https://docs.microsoft.com/en-us/vsts/integrate/previous-apis/build/definitions?view=tfs-2015 ).

I have tested and it works for me.
That way, we don't need any more to do this query (so neither my code nor the one in your PR). Do you think we should remove this call or keep it as fallback?

Reply:
Good idea! It should work and be sufficient, no need to keep a fallback, as team projects are unique for a given instance.

We should just be careful to properly escape the segment (a team project "foo bar" needs to result in a foo%20bar url segment). Yes, my VSTS test case is insane 😃 

The [other PR](https://github.com/pmiossec/gitextensions/pull/3) opened yesterday has been updated to do just that. It actually makes things way more simple, as the project name can be stashed in the BaseAddress.